src/Parsers/IAST.h

```diff
@@ -181,29 +181,84 @@ class IAST : public std::enable_shared_from_this<IAST>, public TypePromotion<IAS
+
+        bool hilite;
+        static const char * hilite_keyword;
+        static const char * hilite_identifier;
+        static const char * hilite_function;
+        static const char * hilite_operator;
+        static const char * hilite_alias;
+        static const char * hilite_substitution;
+        static const char * hilite_none;
+        static const char * hilite_metacharacter;
+
+        /**
+         * To be used in RAII style, just like std::lock_guard with std::mutex.
+         */
+        class Hiliter
         {
-            nl_or_ws = one_line ? ' ' : '\n';
-        }
+            friend struct FormatSettings;
+            DB::WriteBuffer & ostr;
+            bool hilite;
+            Hiliter(DB::WriteBuffer & ostr_, bool hilite_, const char * hilite_type);
+            Hiliter(const Hiliter & other) = delete;
+            Hiliter(Hiliter && other) = delete;
+            Hiliter & operator=(const Hiliter & other) = delete;
+            Hiliter & operator=(Hiliter && other) = delete;
+            ~Hiliter();
+        };
+        Hiliter createHiliter(const char * hilite_type) const;
+        void writePossiblyHilited(std::string_view str, const char * hilite_type) const;
+        void writeIdentifierOrAlias(const String & name, bool should_hilite_as_alias = false) const;
+
+        bool should_show_secrets; /// Show secret parts of the AST (e.g. passwords, encryption keys).
+    public:
+        FormatSettings(WriteBuffer & ostr_, bool one_line_, bool always_quote_identifiers_ = false,
+                       IdentifierQuotingStyle identifier_quoting_style_ = IdentifierQuotingStyle::Backticks, bool hilite_ = false,
+                       bool should_show_secrets_ = true)
+            : ostr(ostr_), one_line(one_line_), always_quote_identifiers(always_quote_identifiers_),
+            identifier_quoting_style(identifier_quoting_style_), hilite(hilite_), should_show_secrets(should_show_secrets_) {}
 
         FormatSettings(WriteBuffer & ostr_, const FormatSettings & other)
-            : ostr(ostr_), hilite(other.hilite), one_line(other.one_line),
-            always_quote_identifiers(other.always_quote_identifiers), identifier_quoting_style(other.identifier_quoting_style)
-        {
-            nl_or_ws = one_line ? ' ' : '\n';
-        }
+            : ostr(ostr_), one_line(other.one_line), always_quote_identifiers(other.always_quote_identifiers),
+            identifier_quoting_style(other.identifier_quoting_style), hilite(other.hilite),
+            should_show_secrets(other.should_show_secrets) {}
+
+        FormatSettings(const FormatSettings & other, bool always_quote_identifiers_)
+            : ostr(other.ostr), one_line(other.one_line), always_quote_identifiers(always_quote_identifiers_),
+            identifier_quoting_style(other.identifier_quoting_style), hilite(other.hilite),
+            should_show_secrets(other.should_show_secrets) {}
+
+        void writeKeyword(std::string_view str) const;
```

Comment:
A thought about this PR a bit. `settings.writeKeyword()` just doesn't feel right. The `FormatSettings` was a simple data structure, and now it's going to become a complex class. The code seems to become more complex and difficult to understand than it was before.

I read how you came to this. But I suggest another approach - let's keep `FormatSettings` simple and move those function to a new class Formatter along with `ostr`.

```
class IAST
{
...
    class Formatter
    {
    public:
        Formatter(WriteBuffer & out, const FormatSettings & settings, FormatState & state);
        Formatter(const Formatter & src); // make Formatter copyable
        void setNeedParent(...)
        void writeFunction(std::string_view name) const;
        ...
    private:
        WriteBuffer & out; // settings.ostr moved here
        const FormatSettings & settings;
        FormatState & state;
        
        // Fields of FormatStateStacked:
        UInt8 indent = 0;
        bool need_parens = false;
        ...
    };
    
    void format(WriteBuffer & out, const FormatSettings & settings) const
    {
        FormatState state;
        formatImpl(Formatter{out, settings, state);
    }
    
    virtual void formatImpl(const Formatter & formatter) const;
};
```

Thus this new class will be instead of FormatStateStacked and it must be copyable - which means for example `ASTFunction::formatImpl` will be able to copy it, change something and call `formatImpl` on arguments with the changed formatter.

Reply:
This is basically taking the initial idea to the next level.

I'm all in for it, as it is quite confusing to have **three** formatting settings in `formatImpl`.

I'll take a look into the other two classes later and will be back.